File: examples/mass_ulw_web.mjs
import { spawn } from "node:child_process";

export const MAX_WEB_WORKERS = 2;
export const SPAWN_STAGGER_SECONDS = 10;

function commonArgs(bridgeUrl) {
  return bridgeUrl ? ["--bridge-url", bridgeUrl] : [];
}

function requireText(value, name) {
  if (typeof value !== "string" || value.trim() === "") {
    throw new Error(`${name} must be non-empty`);
  }
  return value.trim();
}

export function singleInvocation({
  task,
  label = "web",
  workspace,
  bridgeUrl,
  binary = "delegate_to_chatgpt_web",
}) {
  requireText(label, "label");
  const args = commonArgs(bridgeUrl);
  if (workspace) args.push("--workspace", workspace);
  args.push("--stdin", "--json");
  return { program: binary, args, stdin: requireText(task, "task") };
}

export function parallelPairInvocation({
  tasks,
  bridgeUrl,
  binary = "delegate_to_chatgpt_web",
}) {
  if (!Array.isArray(tasks) || tasks.length !== MAX_WEB_WORKERS) {
    throw new Error(`parallelPairInvocation requires exactly ${MAX_WEB_WORKERS} tasks`);
  }
  const normalized = tasks.map((item) => ({
    label: requireText(item.label, "label"),
    task: requireText(item.task, "task"),
    ...(item.workspace ? { workspace: item.workspace } : {}),
  }));
  return {
    program: binary,
    args: [...commonArgs(bridgeUrl), "--batch-stdin", "--json"],
    stdin: JSON.stringify({ tasks: normalized }),
  };
}

export function resumeInvocation({
  scopeId,
  task,
  bridgeUrl,
  binary = "delegate_to_chatgpt_web",
}) {
  return {
    program: binary,
    args: [
      ...commonArgs(bridgeUrl),
      "--resume-scope",
      requireText(scopeId, "scopeId"),
      "--stdin",
      "--json",
    ],
    stdin: requireText(task, "task"),
  };
}

export function closeInvocation({
  scopeId,
  bridgeUrl,
  binary = "delegate_to_chatgpt_web",
}) {
  return {
    program: binary,
    args: [
      ...commonArgs(bridgeUrl),
      "--close-scope",
      requireText(scopeId, "scopeId"),
      "--json",
    ],
  };
}

export async function runInvocation(invocation, { cwd, env = process.env } = {}) {
  return await new Promise((resolve, reject) => {
    const child = spawn(invocation.program, invocation.args, {
      cwd,
      env,
      stdio: ["pipe", "pipe", "pipe"],
    });
    let stdout = "";
    let stderr = "";
    child.stdout.setEncoding("utf8");
    child.stderr.setEncoding("utf8");
    child.stdout.on("data", (chunk) => (stdout += chunk));
    child.stderr.on("data", (chunk) => (stderr += chunk));
    child.on("error", reject);
    child.on("close", (code) => {
      if (code !== 0) {
        reject(new Error(`delegate_to_chatgpt_web exited ${code}: ${stderr.trim()}`));
        return;
      }
      try {
        resolve(JSON.parse(stdout));
      } catch (error) {
        reject(new Error(`invalid delegate JSON: ${error.message}; stdout=${stdout}`));
      }
    });
    if (invocation.stdin !== undefined) child.stdin.write(invocation.stdin);
    child.stdin.end();
  });
}

export function retainedScope(result, label) {
  if (!result?.terminal) throw new Error("delegate result is not terminal");
  const item = result.delegations?.find((entry) => entry.label === label);
  if (!item) throw new Error(`missing delegation label: ${label}`);
  if (!item.session_retained || !item.resumable || !item.scope_id) {
    throw new Error(`delegation ${label} is not retained/resumable`);
  }
  return item.scope_id;
}

export function fanInPrompt(result, integrationGoal, localVerificationFeedback = "") {
  if (!result?.terminal) throw new Error("delegate result is not terminal");
  const evidence = (result.delegations ?? [])
    .map(
      (item) =>
        `- ${item.label ?? "unlabeled"}: terminal=${item.terminal_state ?? "UNKNOWN"}; detail=${item.terminal_detail ?? "none"}`,
    )
    .join("\n");
  const feedback = localVerificationFeedback.trim() || "No additional local verification feedback was supplied.";
  return `Fan-in integration pass. Both parallel Web workers are terminal. Continue in this same retained session; do not create another Web worker.\n\nIntegration goal:\n${integrationGoal.trim()}\n\nParallel terminal evidence:\n${evidence}\n\nLocal verification feedback:\n${feedback}\n\nInspect the current workspace state independently, reconcile both domains, fix integration defects, run the required local verification, and finish only after authoritative completion_check is ready=true.`;
}

File: examples/mass_ulw_web.py
"""Python eval helper for mass-ulw -> delegate_to_chatgpt_web workflows."""

from __future__ import annotations

from dataclasses import dataclass
import json
import os
import subprocess
from typing import Any, Mapping, Sequence

MAX_WEB_WORKERS = 2
SPAWN_STAGGER_SECONDS = 10


@dataclass(frozen=True)
class Invocation:
    program: str
    args: tuple[str, ...]
    stdin: str | None = None


def _text(value: str, name: str) -> str:
    value = value.strip()
    if not value:
        raise ValueError(f"{name} must be non-empty")
    return value


def _common_args(bridge_url: str | None) -> list[str]:
    return ["--bridge-url", bridge_url] if bridge_url else []


def single_invocation(
    *,
    task: str,
    label: str = "web",
    workspace: str | None = None,
    bridge_url: str | None = None,
    binary: str = "delegate_to_chatgpt_web",
) -> Invocation:
    _text(label, "label")
    args = _common_args(bridge_url)
    if workspace:
        args.extend(["--workspace", workspace])
    args.extend(["--stdin", "--json"])
    return Invocation(binary, tuple(args), _text(task, "task"))


def parallel_pair_invocation(
    *,
    tasks: Sequence[Mapping[str, str]],
    bridge_url: str | None = None,
    binary: str = "delegate_to_chatgpt_web",
) -> Invocation:
    if len(tasks) != MAX_WEB_WORKERS:
        raise ValueError(f"parallel_pair_invocation requires exactly {MAX_WEB_WORKERS} tasks")
    normalized: list[dict[str, str]] = []
    for item in tasks:
        entry = {
            "label": _text(item["label"], "label"),
            "task": _text(item["task"], "task"),
        }
        if item.get("workspace"):
            entry["workspace"] = item["workspace"]
        normalized.append(entry)
    args = [*_common_args(bridge_url), "--batch-stdin", "--json"]
    return Invocation(binary, tuple(args), json.dumps({"tasks": normalized}))


def resume_invocation(
    *,
    scope_id: str,
    task: str,
    bridge_url: str | None = None,
    binary: str = "delegate_to_chatgpt_web",
) -> Invocation:
    args = [
        *_common_args(bridge_url),
        "--resume-scope",
        _text(scope_id, "scope_id"),
        "--stdin",
        "--json",
    ]
    return Invocation(binary, tuple(args), _text(task, "task"))


def close_invocation(
    *,
    scope_id: str,
    bridge_url: str | None = None,
    binary: str = "delegate_to_chatgpt_web",
) -> Invocation:
    args = [
        *_common_args(bridge_url),
        "--close-scope",
        _text(scope_id, "scope_id"),
        "--json",
    ]
    return Invocation(binary, tuple(args))


def run_invocation(
    invocation: Invocation,
    *,
    cwd: str | None = None,
    env: Mapping[str, str] | None = None,
) -> dict[str, Any]:
    completed = subprocess.run(
        [invocation.program, *invocation.args],
        input=invocation.stdin,
        text=True,
        capture_output=True,
        cwd=cwd,
        env=dict(env) if env is not None else os.environ.copy(),
        check=False,
    )
    if completed.returncode != 0:
        raise RuntimeError(
            f"delegate_to_chatgpt_web exited {completed.returncode}: {completed.stderr.strip()}"
        )
    try:
        return json.loads(completed.stdout)
    except json.JSONDecodeError as error:
        raise RuntimeError(
            f"invalid delegate JSON: {error}; stdout={completed.stdout}"
        ) from error


def retained_scope(result: Mapping[str, Any], label: str) -> str:
    if not result.get("terminal"):
        raise ValueError("delegate result is not terminal")
    for item in result.get("delegations", []):
        if item.get("label") == label:
            if item.get("session_retained") and item.get("resumable") and item.get("scope_id"):
                return str(item["scope_id"])
            raise ValueError(f"delegation {label} is not retained/resumable")
    raise ValueError(f"missing delegation label: {label}")


def fan_in_prompt(
    result: Mapping[str, Any],
    integration_goal: str,
    local_verification_feedback: str = "",
) -> str:
    if not result.get("terminal"):
        raise ValueError("delegate result is not terminal")
    evidence = "\n".join(
        f"- {item.get('label', 'unlabeled')}: terminal={item.get('terminal_state', 'UNKNOWN')}; "
        f"detail={item.get('terminal_detail', 'none')}"
        for item in result.get("delegations", [])
    )
    feedback = (
        local_verification_feedback.strip()
        or "No additional local verification feedback was supplied."
    )
    return (
        "Fan-in integration pass. Both parallel Web workers are terminal. Continue in this same "
        "retained session; do not create another Web worker.\n\n"
        f"Integration goal:\n{integration_goal.strip()}\n\n"
        f"Parallel terminal evidence:\n{evidence}\n\n"
        f"Local verification feedback:\n{feedback}\n\n"
        "Inspect the current workspace state independently, reconcile both domains, fix integration "
        "defects, run the required local verification, and finish only after authoritative "
        "completion_check is ready=true."
    )
